src/datumaro/plugins/data_formats/datumaro/base.py

```diff
@@ -326,3 +286,173 @@ def _load_annotations(item):
+
+    def _load_items(self, parsed) -> List:
+        return []
+
+
+class DatumaroBase(SubsetBase):
+    LEGACY_VERSION = "legacy"
+    CURRENT_DATUMARO_FORMAT_VERSION = DATUMARO_FORMAT_VERSION
+
+    # If Datumaro format version goes up, it will be
+    # ALLOWED_VERSIONS = {LEGACY_VERSION, 1.0, ..., CURRENT_DATUMARO_FORMAT_VERSION}
+    ALLOWED_VERSIONS = {LEGACY_VERSION, CURRENT_DATUMARO_FORMAT_VERSION}
+
+    def __init__(
+        self,
+        path: str,
+        *,
+        subset: Optional[str] = None,
+        stream: bool = True,
+        ctx: Optional[ImportContext] = None,
+    ):
+        assert osp.isfile(path), path
+        subset = osp.splitext(osp.basename(path))[0] if subset is None else subset
+        self._stream = stream
+        super().__init__(subset=subset, ctx=ctx)
+        self._init_path(path)
+
+        dm_version = self._get_dm_format_version(path)
+
+        # when backward compatibility happen, we should implement version specific readers
+        if dm_version not in self.ALLOWED_VERSIONS:
+            raise DatasetImportError(
+                f"Datumaro format version of the given dataset is {dm_version}, "
+                f"but not supported by this Datumaro version: {__version__}. "
+                f"The allowed datumaro format versions are {self.ALLOWED_VERSIONS}. "
+                "Please install the latest Datumaro."
+            )
+
+        self._load_impl(path)
+
+    def _init_path(self, path: str):
+        """Initialize path variables"""
+        rootpath = ""
+        if path.endswith(osp.join(DatumaroPath.ANNOTATIONS_DIR, osp.basename(path))):
+            rootpath = path.rsplit(DatumaroPath.ANNOTATIONS_DIR, maxsplit=1)[0]
+        self._rootpath = rootpath
+
+        images_dir = ""
+        if rootpath and osp.isdir(osp.join(rootpath, DatumaroPath.IMAGES_DIR)):
+            images_dir = osp.join(rootpath, DatumaroPath.IMAGES_DIR)
+        self._images_dir = images_dir
+
+        pcd_dir = ""
+        if rootpath and osp.isdir(osp.join(rootpath, DatumaroPath.PCD_DIR)):
+            pcd_dir = osp.join(rootpath, DatumaroPath.PCD_DIR)
+        self._pcd_dir = pcd_dir
+
+    @property
+    def is_stream(self) -> bool:
+        return self._stream
+
+    def infos(self):
+        return self._reader.infos
+
+    def categories(self):
+        return self._reader.categories
+
+    def media_type(self):
+        return self._reader.media_type
+
+    def _load_impl(self, path: str) -> None:
+        """Actual implementation of loading Datumaro format."""
+        self._reader = (
+            DefaultReader(path, self._subset, self._rootpath, self._images_dir, self._pcd_dir)
+            if not self._stream
+            else StreamDefaultReader(
+                path, self._subset, self._rootpath, self._images_dir, self._pcd_dir
+            )
+        )
+        return self._reader
+
+    def _get_dm_format_version(self, path) -> str:
```

Comment:
Why don't you put this into `DefaultReader` too?

Reply:
This should be done before going to `reader`. This is because we choose an appropriate `reader` according to `dm_format_version` in the future.

Reply:
Can I ask you why this should be done before going to `reader`?